### ISSUE TITLE
SSH Websocket: bad handshake

### ISSUE THREAD

**Issue by ewu00** [cloudflare/cloudflared#404]
I'm having a similar issue as #364 but with SSH access. I'm able to connect to RDP just fine via cloudflared proxy. SSH doesn't work through cloudflared proxy nor via web rendering. SSH on server side to 'localhost' works fine. I've turned on debug logs for cloudflared and ssh on server side but there is no output during SSH connection attempts at all. Seems I'm caught in some kind of auth loop?

Server: Ubuntu 18.04.5
Client: Kubuntu 20.04.2

Config:
```
tunnel: <redacted>
credentials-file: /root/.cloudflared/<redacted>.json
transport-loglevel: debug
logfile: /tmp/cloudflared.log

ingress:
  - hostname: cf-tunnel-test.<redacted>
    service: rdp://localhost:3389
  - hostname: cf-tunnel-test-ssh.<redacted>
    service: ssh://localhost:22
  - service: http_status:404
```

Client debug output:
```
$ ssh cf-tunnel-test-ssh.<redacted>        
2021-06-22T21:17:23Z DBG Websocket request: GET / HTTP/1.1
Host: cf-tunnel-test-ssh.<redacted>


2021-06-22T21:17:24Z DBG Access Websocket request: GET / HTTP/1.1
Host: cf-tunnel-test-ssh.<redacted>
Cf-Access-Token: <redacted>


2021-06-22T21:17:24Z DBG Websocket response: "HTTP/1.1 302 Moved Temporarily\r\nTransfer-Encoding: chunked\r\nAccess-Control-Allow-Credentials: true\r\nAlt-Svc: h3-27=\":443\"; ma=86400, h3-28=\":443\"; ma=86400, h3-29=\":443\"; ma=86400, h3=\":443\"; ma=86400\r\nCache-Control: private, max-age=0, no-store, no-cache, must-revalidate, post-check=0, pre-check=0\r\nCf-Ray: 663880d33c1c3ff7-YYZ\r\nCf-Request-Id: 0ad72ed80500003ff75639a000000001\r\nConnection: keep-alive\r\nDate: Tue, 22 Jun 2021 21:17:24 GMT\r\nExpect-Ct: max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"\r\nExpires: Thu, 01 Jan 1970 00:00:01 GMT\r\nLocation: https://<redacted>.cloudflareaccess.com/cdn-cgi/access/login/<redacted>&redirect_url=%2F&meta=<redacted>\r\nNel: {\"report_to\":\"cf-nel\",\"max_age\":604800}\r\nReport-To: {\"endpoints\":[{\"url\":\"https:\\/\\/a.nel.cloudflare.com\\/report\\/v2?s=<redacted>%3D\"}],\"group\":\"cf-nel\",\"max_age\":604800}\r\nServer: cloudflare\r\n\r\n0\r\n\r\n"
2021-06-22T21:17:25Z DBG Access Websocket request: GET / HTTP/1.1
Host: cf-tunnel-test-ssh.<redacted>
Cf-Access-Token: <redacted>


2021-06-22T21:17:25Z DBG Websocket response: "HTTP/1.1 302 Moved Temporarily\r\nTransfer-Encoding: chunked\r\nAccess-Control-Allow-Credentials: true\r\nAlt-Svc: h3-27=\":443\"; ma=86400, h3-28=\":443\"; ma=86400, h3-29=\":443\"; ma=86400, h3=\":443\"; ma=86400\r\nCache-Control: private, max-age=0, no-store, no-cache, must-revalidate, post-check=0, pre-check=0\r\nCf-Ray: 663880d93a443fd2-YYZ\r\nCf-Request-Id: 0ad72edbc300003fd2562e5000000001\r\nConnection: keep-alive\r\nDate: Tue, 22 Jun 2021 21:17:25 GMT\r\nExpect-Ct: max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"\r\nExpires: Thu, 01 Jan 1970 00:00:01 GMT\r\nLocation: https://<redacted>.cloudflareaccess.com/cdn-cgi/access/login/cf-tunnel-test-ssh.<redacted>?kid=<redacted>\r\nNel: {\"report_to\":\"cf-nel\",\"max_age\":604800}\r\nReport-To: {\"endpoints\":[{\"url\":\"https:\\/\\/a.nel.cloudflare.com\\/report\\/v2?s=%2F<redacted>%3D\"}],\"group\":\"cf-nel\",\"max_age\":604800}\r\nServer: cloudflare\r\n\r\n0\r\n\r\n"
2021-06-22T21:17:25Z ERR failed to connect to origin error="websocket: bad handshake" originURL=https://<redacted>
websocket: bad handshake
kex_exchange_identification: Connection closed by remote host
```

**Comment by nmldiegues** [comment]
Can you share the SSH config?

Does it work if you temporarily remove the Access policy?

**Comment by ewu00** [comment]
If I remove access policy, I'm not allowed to access. If I add access policy (google as authentication and set email domain in include), then I get prompted to login in the browser and then says "Success! A token has been returned to the machine that initiated the request. Feel free to close this browser window." but I get the websocket error after. Again, this works fine for RDP proxying.

cloudflared version (.deb package): 2021.6.0 (built 2021-06-22-0801 UTC)

SSH client config:
```
Host *.<redacted>
  ProxyCommand /usr/local/bin/cloudflared access ssh --hostname %h --loglevel debug
```

SSH server config is default from 18.04 with exception of enabling DEBUG log level:
```
LogLevel DEBUG
ChallengeResponseAuthentication no
UsePAM yes
X11Forwarding yes
PrintMotd no
AcceptEnv LANG LC_*
Subsystem       sftp    /usr/lib/openssh/sftp-server
```

**Comment by nmldiegues** [comment]
I meant removing the Access rule in the sense that there'd be no Access protection to that origin. Just to try to understand where the problem could be coming from.

Have you enabled the browser rendering as per https://developers.cloudflare.com/cloudflare-one/tutorials/ssh-browser to try this without the SSH client and using just the browser?

**Comment by ewu00** [comment]
I'm not sure what you mean by "Access protection" then. I removed access policies in Cloudflare for Teams application. Then I wasn't permitted to access. Is there something else you're thinking of?

Yes, I've tried browser rendering and it says can't connect. If I use developer tools in chrome I saw websocket failure and also the same 302 redirect loop

**Comment by nmldiegues** [comment]
> I'm not sure what you mean by "Access protection" then. I removed access policies in Cloudflare for Teams application. Then I wasn't permitted to access.

This is what's not clear to me. The hostname that you are using is public. If you remove the Cloudflare for Teams Access application, and thus should no longer get an "Access prompt login", it should get you direct access to the Tunnel origin (and then you have to auth to the origin, be it SSH, RDP, etc...).

What do you mean with "wasn't permitted to access"?

**Comment by ewu00** [comment]
The hostname is public but it points to a cloudflared tunnel and the SSH service is not exposed publicly by design. This is per documentation here: https://developers.cloudflare.com/cloudflare-one/tutorials/ssh

**Comment by nmldiegues** [comment]
Correct, I understand that. I was trying to understand if you skip "Create a Zero Trust policy" and use "Native Terminal", then you should not get an Access login redirect, and whether the problem prevailed or not.

**Comment by ewu00** [comment]
Oh, I see. I didn't realize that I could skip the Zero Trust policy to still use cloudflare proxy. Yes, this works fine so the tunnel is fine.

I found the issue. In the docs below, the image has "Enable Binding Cookie" enabled and its description says it's for security so naturally I enabled it too. However, this setting is causing my issues here. As it turns out, it was the one difference vs the RDP application I created.
https://developers.cloudflare.com/cloudflare-one/tutorials/ssh-browser#connect-from-a-browser-rendered-terminal
https://developers.cloudflare.com/cloudflare-one/tutorials/ssh#browser-rendered-terminal

![image](https://user-images.githubusercontent.com/84043724/123128767-6a711980-d419-11eb-9899-e8913a4562a6.png)

Looking up this option now, I see that it can only be used for HTTP applications: https://developers.cloudflare.com/cloudflare-one/identity/users/session-management#when-not-to-use-the-binding-cookie

Maybe the guide should have an updated image and a note on this?
